IMX95-M7 Bugfix for TPM3 not being able to control the clock

PetervdPerk-NXP opened the pull request:
Fixes following error on IMX95-M7 when using TPM3.
```
[00:00:00.000,000] <err> pwm_mcux_tpm: Could not turn on clock
```
TPM3 is clocked by the BUSWAKEUP instead of having it's own specific TPM clock.
BUSWAKEUP is typically fixed and can't be controlled.

On IMX9596 the AON/WU clocks aren't controllable are on by default. But TPM1 and TPM3 uses the clocks, by not enabling these clocks  on both TPM1/TPM3 will initialize successfully.

henrikbrixandersen reviewed (changes requested):
We already have means for representing a fixed-clock in the devicetree: https://docs.zephyrproject.org/latest/build/dts/api/bindings/clock/fixed-clock.html

Why not use that as clock source?

PetervdPerk-NXP commented:
> We already have means for representing a fixed-clock in the devicetree: https://docs.zephyrproject.org/latest/build/dts/api/bindings/clock/fixed-clock.html
> 
> Why not use that as clock source?

Because we still need to able to fetch the rate of the clock.
https://github.com/zephyrproject-rtos/zephyr/blob/1590972ce363910ffbcbf4b4c8beb350e38ee11d/drivers/pwm/pwm_mcux_tpm.c#L162
This rate of the clock be different due to how the boot core (CM33) configures this clock but we can't control this clock.

Alternatively we don't let `clock_control_on` fail if it's already on but that requires non-portable changes in `clock_control_arm_scmi`.

@JiafeiPan what would be your preference or do think there's another alternative?

henrikbrixandersen commented:
> > We already have means for representing a fixed-clock in the devicetree: https://docs.zephyrproject.org/latest/build/dts/api/bindings/clock/fixed-clock.html
> > Why not use that as clock source?
> 
> Because we still need to able to fetch the rate of the clock.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/1590972ce363910ffbcbf4b4c8beb350e38ee11d/drivers/pwm/pwm_mcux_tpm.c#L162
> 
> 
> This rate of the clock be different due to how the boot core (CM33) configures this clock but we can't control this clock.
> Alternatively we don't let `clock_control_on` fail if it's already on but that requires non-portable changes in `clock_control_arm_scmi`.
> 
> @JiafeiPan what would be your preference or do think there's another alternative?

Ah, got it. I'd recommend using another term than "fixed clock" then for the property, as that already has a well-established meaning. This is an "always on" clock, as I understand it, not a "fixed (frequency)" clock.

henrikbrixandersen commented:
If this is correct, then one solution could be for the clock controller driver to ignore requests for turning this clock on/off and free the driver from this concern.

PetervdPerk-NXP commented:
> Ah, got it. I'd recommend using another term than "fixed clock" then for the property, as that already has a well-established meaning. This is an "always on" clock, as I understand it, not a "fixed (frequency)" clock.

Good point, `fixed-clock` was confusing, I've renamed it to `always-on-clock`.

henrikbrixandersen commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -94,6 +94,7 @@
 			prescaler = <16>;
 			status = "disabled";
 			#pwm-cells = <3>;
+			always-on-clock;
```
Why not simply have the clock controller handling the `IMX95_CLK_BUSWAKEUP` clock treat `clock_control_on()` calls gracefully?

PetervdPerk-NXP replied:
The clock controller is abstracted away by SCMI and thus doesn't have a notion of `IMX95_CLK_BUSWAKEUP`
https://github.com/zephyrproject-rtos/zephyr/blob/7fb92c9eb125f722da96a9df56622bd0b745f9eb/drivers/firmware/scmi/clk.c

I suppose for a board specific solution something could work would be to define a `fixed-clock` with a predefined frequency so that the `kinetis-tpm` driver wouldn't fail on initialization. 

The approach in this PR kept `clock_control_get_rate` API intact thus avoiding having predefined frequencies but also at the cost adding the dts entry/complexity.

JiafeiPan commented:
> > We already have means for representing a fixed-clock in the devicetree: https://docs.zephyrproject.org/latest/build/dts/api/bindings/clock/fixed-clock.html
> > Why not use that as clock source?
> 
> Because we still need to able to fetch the rate of the clock.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/1590972ce363910ffbcbf4b4c8beb350e38ee11d/drivers/pwm/pwm_mcux_tpm.c#L162
> 
> 
> This rate of the clock be different due to how the boot core (CM33) configures this clock but we can't control this clock.
> Alternatively we don't let `clock_control_on` fail if it's already on but that requires non-portable changes in `clock_control_arm_scmi`.
> 
> @JiafeiPan what would be your preference or do think there's another alternative?

Sorry for the late response. Actually as CM33 have no permission to control this clock, it will return SCMI_DENIED to indicate no permission.  so maybe another approach could be checking return value of `clock_control_on(config->clock_dev, config->clock_subsys))`, if the platform is imx95 and also return the value is SCMI_DENIED, we could allow the driver continue to run because this is not an error in this case,  this method only need to update pwm driver.

PetervdPerk-NXP committed:
kinetis-tpm: don't enable AON/WU clocks on IMX9596

On IMX9596 the AON/WU clocks aren't controllable are on by default.
But TPM1 and TPM3 uses the clocks, by not enabling these clocks
on both TPM1/TPM3 will initialize succesfully.

Signed-off-by: Peter van der Perk <peter.vanderperk@nxp.com>

PetervdPerk-NXP commented:
@henrikbrixandersen @JiafeiPan 
I made an ifdef for the `SOC_MIMX9596` where it doesn't execute `clock_control_on` for both `CLK_BUSWAKEUP` and `CLK_BUSAON`. This is working on my IMX95-EVK board.

You guys okay with this solution?

PetervdPerk-NXP commented on drivers/pwm/pwm_mcux_tpm.c:
```diff
@@ -151,10 +155,18 @@ static int mcux_tpm_init(const struct device *dev)
+	if (config->clock_subsys != (clock_control_subsys_t)IMX95_CLK_BUSWAKEUP &&
+	    config->clock_subsys != (clock_control_subsys_t)IMX95_CLK_BUSAON) {
+#endif
+		if (clock_control_on(config->clock_dev, config->clock_subsys)) {
```
SCMI_DENI would be a problem for TPM4 if the system-manager doesn't give you permissions to the TPM4_CLOCK.

JiafeiPan replied:
okay, then I am okay with you current solution.

bperseghetti commented:
I think this just needs a review by @anangl? Or @dleach02 does this count as an insular "NXP specific" commit since it only touches pwm_mcux?
